### PR TITLE
Add unit tests for kaniko build in odo deploy

### PR DESCRIPTION
**Pull request by jaideepr97**
**What type of PR is this?**
> /kind code-refactoring

**What does does this PR do / why we need it**:
- Adds unit tests for newly added code for kaniko build in ```odo deploy```
- Adds cleanup of builder pod after build is completed
- reorganizes code to avoid duplication of logic for creation of secret 

**How to test changes / Special notes to the reviewer**:
- create a new directory, and run ```odo create nodejs --starter=nodejs-starter```
- run ```odo url create```
- replace created devfile with the following sample and change ```Rootless``` depending on what kind of build you want to perform
```
schemaVersion: 2.1.0
metadata:
  name: nodejs
  version: 1.0.0
  alpha.deployment-manifest: "https://raw.githubusercontent.com/groeges/devfile-registry/master/devfiles/nodejs/deploy_deployment.yaml"
projects:
  - name: nodejs-starter
    git:
      location: "https://github.com/odo-devfiles/nodejs-ex.git"
components:
  - container:
      name: runtime
      image: registry.access.redhat.com/ubi8/nodejs-12:1-36
      memoryLimit: 1024Mi
      mountSources: true
      endpoints:
        - name: http-3000
          targetPort: 3000
          configuration:
            protocol: tcp
            scheme: http
            type: terminal

  - dockerfile:
      name: dockerfile-build
      source: 
         sourceDir: "src"
         location: "https://github.com/ranakan19/golang-ex.git"
      dockerfileLocation: "https://raw.githubusercontent.com/jaideepr97/OpenShift-guides/master/Dockerfile"
      destination:
      Rootless: true 
    
commands:
  - exec:
      id: install
      component: runtime
      commandLine: npm install
      workingDir: /project
      group:
        kind: build
        isDefault: true
  - exec:
      id: run
      component: runtime
      commandLine: npm start
      workingDir: /project
      group:
        kind: run
        isDefault: true
  - exec:
      id: debug
      component: runtime
      commandLine: npm run debug
      workingDir: /project
      group:
        kind: debug
        isDefault: true
  - exec:
      id: test
      component: runtime
      commandLine: npm test
      workingDir: /project
      group:
        kind: test
        isDefault: true
```
- run ```odo deploy --tag=<imageTag> ```

### CODE REVIEW

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 1]
```diff
@@ -211,9 +211,14 @@ func (a Adapter) Build(parameters common.BuildParameters) (err error) {
...
-		if err := a.createDockerConfigSecret(parameters); err != nil {
+		dockerConfigSecretBytes, err := a.createDockerConfigSecretBytes(parameters.DockerConfigJSONFilename)
+		if err != nil {
 			return err
 		}
+
+		if _, err := a.createDockerConfigSecret(regcredName, parameters.EnvSpecificInfo.GetNamespace(), dockerConfigSecretBytes); err != nil {
+			return errors.Wrap(err, "failed to create dockerconfig secret")
+		}
 	}
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter.go:218]: The original code is definitely cleaner.   I would leave it as a.createDockerConfigSecret()

File: pkg/devfile/adapters/kubernetes/utils/utils.go [hunk 2]
```diff
@@ -361,21 +362,32 @@ func PipeStdOutput(cmdOutput string, reader *io.PipeReader, controlC chan os.Sig
...
-	if err != nil {
-		return nil, fmt.Errorf("failed to read Docker config %#v : %s", filename, err)
+	if fs != nil {
+		f, err := fs.Open(filename)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read test Docker config %#v : %s", filename, err)
+		}
+		filePointer = f
+		defer f.Close()
+	} else {
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/utils/utils.go]: Why introduce the complexity here to this rather simple straightforward code?

Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/utils/utils.go]: I was trying to write test cases for this function using an afero FS object but had no way to pass the test filesystem object to the function so I had to restructure it to accept the filesystem object so I could run the test

File: pkg/devfile/adapters/kubernetes/component/adapter.go (hunk 1 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter.go:218]: would it be better if I renamed ```createDockerConfigSecretBytes()``` to something else, and added a call to  ```createDockerConfigSecret()``` in that function at the end? (Same in kaniko_build.go)
That way we can have it be called something similar to ```createDockerConfigSecret``` while reusing the create secret logic 
and there will only be 1 high level function call for creating docker secret

Comment (wtam2018) [reply pkg/devfile/adapters/kubernetes/component/adapter.go:218]: It is not a function naming issue.   The caller does not care about the byte array of the secret.   The caller only cares about getting secrete created on the cluster.    The unit test should be validating that a.createDockerConfigSecret() has successfully created an expected secret if I provide a docker config file path in the parameters.

File: pkg/devfile/adapters/kubernetes/utils/utils.go (hunk 2 above)
Comment (wtam2018) [reply pkg/devfile/adapters/kubernetes/utils/utils.go]: If you are going to pass afero fs and use it to do file i/o, we should make the normal call flow (not testing code) to use afero fs as well.

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go [hunk 3]
```diff
@@ -555,6 +569,148 @@ func TestAdapterDelete(t *testing.T) {
 	}
 }
 
+func TestCreateDockerConfigSecretBytes(t *testing.T) {
+
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter_test.go]: This test is combined (covered by) `TestCreateDockerConfigSecret() that tests the "end-to-end' of creating a docker config secret from a file path input.

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go [hunk 4]
```diff
@@ -555,6 +569,148 @@ func TestAdapterDelete(t *testing.T) {
...
+		t.Errorf("failed to retrieve dockerconfig secret bytes")
+	}
+
+	if !reflect.DeepEqual(want, got) {
+		t.Errorf("secret bytes don't match")
+	}
+}
+
+func TestCreateDockerConfigSecret(t *testing.T) {
+	testConfigJsonString := "{ \"auths\" : { \"https://index.docker.io/v1/\": { \"auth\": \"test-auth-token\", \"email\": \"test-email\"} },\"HttpHeaders\": {	\"User-Agent\": \"Docker-Client/19.03.8 (darwin)\"},\"experimental\": \"disabled\"}"
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter_test.go:573]: Create a test-data/config.json and pass the file path (relative path "test-data/config.son") to the Parameters which is input to createDockerConfigSecret()

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go [hunk 5]
```diff
@@ -555,6 +569,148 @@ func TestAdapterDelete(t *testing.T) {
...
+		Client: *fkclient,
+	}
+
+	got, err := testAdapter.createDockerConfigSecret(testSecretName, testNs, testDockerConfigData)
+
+	if err != nil {
+		t.Error(err)
+		t.Errorf("failed to get secret")
+	}
+
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter_test.go:638]: Call to the fakeclient to retrive docker secret and validate it (this is the "got").

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go [hunk 6]
```diff
@@ -555,6 +569,148 @@ func TestAdapterDelete(t *testing.T) {
...
+	if !reflect.DeepEqual(got, want) {
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter_test.go]: Use `cmp.Diff` instead.    The output from cmp.Diff can be reported and used for troubleshooting.  It return "" if the two objects are identical.

File: pkg/devfile/adapters/kubernetes/component/kaniko_build.go [hunk 7]
```diff
@@ -67,6 +68,15 @@ func (a Adapter) runKaniko(parameters common.BuildParameters, isImageRegistryInt
 		return errors.Wrapf(err, "error while waiting for pod %s", podSelector)
 	}
 
+	defer func() {
+		// This will clean up the builder pod after build is complete
+		derr := a.Client.KubeClient.CoreV1().Pods(pod.Namespace).Delete(pod.Name, &metav1.DeleteOptions{})
+
+		if err == nil {
+			err = derr
+		}
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/kaniko_build.go]:
```suggestion
		if err := a.Client.KubeClient.CoreV1().Pods(pod.Namespace).Delete(pod.Name, &metav1.DeleteOptions{}); err != nil {
			log.Errorf("Failed to delete pod '%s': %v", pod.Name, err)
		}
```

**Review by wtam2018 — COMMENTED** [review]
I would leave non-testing codes intact as possible.

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go (hunk 3 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter_test.go]: removed this test

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go (hunk 4 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter_test.go:573]: Done

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go (hunk 5 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter_test.go:638]: Done

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go (hunk 6 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter_test.go]: Done

File: pkg/devfile/adapters/kubernetes/component/kaniko_build.go (hunk 7 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/kaniko_build.go]: Done

File: pkg/devfile/adapters/kubernetes/utils/utils.go (hunk 2 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/utils/utils.go]: Removed use of afero fs, just using os now

File: pkg/devfile/adapters/kubernetes/component/adapter.go (hunk 1 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter.go:218]: reverted to old version

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 8]
```diff
@@ -907,19 +908,22 @@ func (a Adapter) Exec(command []string) error {
 	return a.ExecuteCommand(componentInfo, command, true, nil, nil)
 }
 
-func (a Adapter) createDockerConfigSecret(parameters common.BuildParameters) error {
-	data, err := utils.CreateDockerConfigDataFromFilepath(parameters.DockerConfigJSONFilename)
+func (a Adapter) createDockerConfigSecret(DockerConfigJSONFilename string, secretName string, nameSpace string) error {
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter.go:911]: Formal parameter name (DockerConfigJSONFilename) should start with lower case (e.g. dockerConfigJSONFilename)

Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter.go]:
```suggestion
func (a Adapter) createDockerConfigSecret(dockerConfigJSONFilename, secretName, namespace string) error {
```

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 9]
```diff
@@ -907,19 +908,22 @@ func (a Adapter) Exec(command []string) error {
...
+
+	dockerConfigSecretBytes, err := utils.CreateDockerConfigDataFromFilepath(DockerConfigJSONFilename)
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter.go]:
```suggestion
	dockerConfigSecretBytes, err := utils.CreateDockerConfigDataFromFilepath(dockerConfigJSONFilename)
```

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 10]
```diff
@@ -907,19 +908,22 @@ func (a Adapter) Exec(command []string) error {
...
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Error retriving docker config secret bytes")
 	}
-	secretUnstructured, err := utils.CreateSecret(regcredName, parameters.EnvSpecificInfo.GetNamespace(), data)
+
+	secretUnstructured, err := utils.CreateSecret(secretName, nameSpace, dockerConfigSecretBytes)
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter.go]:
```suggestion
	secretUnstructured, err := utils.CreateSecret(secretName, namespace, dockerConfigSecretBytes)
```

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 11]
```diff
@@ -907,19 +908,22 @@ func (a Adapter) Exec(command []string) error {
...
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to convert created secret to unstructured format")
 	}
-	if _, err := a.Client.DynamicClient.Resource(secretGroupVersionResource).
-		Namespace(parameters.EnvSpecificInfo.GetNamespace()).
-		Create(secretUnstructured, metav1.CreateOptions{}); err != nil {
-		return err
+	_, err = a.Client.DynamicClient.Resource(secretGroupVersionResource).
+		Namespace(nameSpace).
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter.go]:
```suggestion
		Namespace(namespace).
```

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go [hunk 12]
```diff
@@ -555,6 +569,128 @@ func TestAdapterDelete(t *testing.T) {
...
+		Type:       corev1.SecretTypeDockerConfigJson,
+		Data: map[string][]byte{
+			corev1.DockerConfigJsonKey: testDockerConfigData,
+		},
+	}
+
+	testSecretData, err := runtimeUnstructured.DefaultUnstructuredConverter.ToUnstructured(testSecret)
+	if err != nil {
+		t.Error(err)
+		t.Errorf("error making map")
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter_test.go]: Delete/combine `	t.Errorf("error making map")`

File: pkg/devfile/adapters/kubernetes/component/kaniko_build.go [hunk 13]
```diff
@@ -131,10 +138,11 @@ func (a Adapter) createDockerCfgSecretForInternalRegistry(ns string) error {
 	// internal registry credentials.
 	secret, err := a.getServiceAccountSecret(ns, builderServiceAccount, corev1.SecretTypeDockercfg)
 	if err != nil {
-		return errors.Wrap(err, "failed to get docker secret")
+		return errors.Wrap(err, "failed to retrieve service account credentials")
 	}
-	if err := a.createDockerConfigSecretFrom(secret); err != nil {
-		return errors.Wrap(err, "failed to create docker secret")
+
+	if err := a.createDockerConfigSecretFrom(secret, secret.GetNamespace()); err != nil {
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/kaniko_build.go]: Don't pass `namespace` as a seperate argument.   We can get it from the `secret` argument.  Don't make it argument unncessarly long.

File: pkg/devfile/adapters/kubernetes/component/kaniko_build_test.go [hunk 14]
```diff
@@ -0,0 +1,455 @@
...
+	if err != nil {
+		t.Errorf("failed to deploy pod using fake client")
+	}
+	got, err := testAdapter.Client.KubeClient.CoreV1().Pods(testAdapter.Client.Namespace).Get("test-component-name-build", metav1.GetOptions{})
+	if err != nil {
+		t.Errorf("failed to deploy pod using fake client")
+	}
+
+	kanikoBuilderPodPorted = *got
+	if !reflect.DeepEqual(kanikoBuilderPodPorted.ObjectMeta, want.ObjectMeta) {
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/kaniko_build_test.go]: Replace with cmp.Diff() calls?

File: pkg/devfile/adapters/kubernetes/utils/utils.go [hunk 15]
```diff
@@ -362,20 +362,21 @@ func PipeStdOutput(cmdOutput string, reader *io.PipeReader, controlC chan os.Sig
 }
 
 func CreateDockerConfigDataFromFilepath(DockerConfigJSONFilename string) ([]byte, error) {
+	var filePointer io.Reader
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/utils/utils.go]:
```suggestion
```

File: pkg/devfile/adapters/kubernetes/utils/utils.go [hunk 16]
```diff
@@ -362,20 +362,21 @@ func PipeStdOutput(cmdOutput string, reader *io.PipeReader, controlC chan os.Sig
...
 	if err != nil {
 		return nil, fmt.Errorf("failed to generate path to file for %s: %v", DockerConfigJSONFilename, err)
 	}
-
 	f, err := os.Open(filename)
 	if err != nil {
 		return nil, fmt.Errorf("failed to read Docker config %#v : %s", filename, err)
 	}
-	data, err := ioutil.ReadAll(f)
+	filePointer = f
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/utils/utils.go]:
```suggestion
```

File: pkg/devfile/adapters/kubernetes/utils/utils.go [hunk 17]
```diff
@@ -362,20 +362,21 @@ func PipeStdOutput(cmdOutput string, reader *io.PipeReader, controlC chan os.Sig
...
+	defer f.Close()
+	data, err := ioutil.ReadAll(filePointer)
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/utils/utils.go]:
```suggestion
	data, err := ioutil.ReadAll(f)
```

File: pkg/devfile/adapters/kubernetes/utils/utils_test.go [hunk 18]
```diff
@@ -620,3 +651,73 @@ func TestUpdateContainersWithSupervisord(t *testing.T) {
...
 }
+
+func TestCreateDockerConfigDataFromFilepath(t *testing.T) {
+	testConfigJsonString := "{ \"auths\" : { \"https://index.docker.io/v1/\": { \"auth\": \"test-auth-token\", \"email\": \"test-email\"} },\"HttpHeaders\": {	\"User-Agent\": \"Docker-Client/19.03.8 (darwin)\"},\"experimental\": \"disabled\"}"
+	testFilename := "test-data/test-config-json"
+
+	got, err := CreateDockerConfigDataFromFilepath(testFilename)
+	if err != nil {
+		t.Error(err)
+		t.Errorf("unable to get dockeconfigdata from filepath")
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/utils/utils_test.go:661]: Having two t.Error in a row seems odd.

Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/utils/utils_test.go:661]: fixed

File: pkg/devfile/adapters/kubernetes/utils/utils.go (hunk 17 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/utils/utils.go]: fixed

File: pkg/devfile/adapters/kubernetes/utils/utils.go (hunk 16 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/utils/utils.go]: fixed

File: pkg/devfile/adapters/kubernetes/utils/utils.go (hunk 15 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/utils/utils.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/kaniko_build_test.go (hunk 14 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/kaniko_build_test.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/kaniko_build.go (hunk 13 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/kaniko_build.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/adapter_test.go (hunk 12 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter_test.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/adapter.go (hunk 11 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/adapter.go (hunk 10 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/adapter.go (hunk 9 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/adapter.go (hunk 8 above)
Comment (jaideepr97) [reply pkg/devfile/adapters/kubernetes/component/adapter.go]: fixed

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 19]
```diff
@@ -907,19 +908,22 @@ func (a Adapter) Exec(command []string) error {
...
+		Create(secretUnstructured, metav1.CreateOptions{})
+	if err != nil {
```
Comment (wtam2018) [inline pkg/devfile/adapters/kubernetes/component/adapter.go]:
```suggestion
	if _, err := a.Client.DynamicClient.Resource(secretGroupVersionResource).
		Namespace(nameSpace).
		Create(secretUnstructured, metav1.CreateOptions{}); err != nil {
```

**Review by wtam2018 — APPROVED** [review]

**Pull request merged**